Add mocked cores and nodes to ActionTesting

markscheel opened the pull request:
## Proposed changes

The ActionTesting framework now can simulate running on multiple nodes and cores.

MockRuntimeSystem now has a `std::vector<size_t>` passed into its constructor, specifying the number of (mock) cores on each (mock) node of the mocked runtime system.  Each `MockDistributedObject` and their proxies keep track of the mock node and core that it lives on. The various member functions in ActionTesting have been updated to do the right thing for multiple nodes and cores.

Test_ActionTesting tests much of the new functionality.

Many of the other tests have been updated to work with the new changes.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
- In ActionTesting, if you are mocking a Group, Nodegroup, or Singleton component, these should now have `chare_type` set to `MockGroupChare`, `MockNodeGroupChare`, or `MockSingletonChare` instead of `MockArrayChare`.  Note that `MockArrayChare` will *sometimes* work for these components for some tests, but not if `ckLocal` or `ckLocalBranch` is ever called on those components.
- In ActionTesting, the new functions `emplace_singleton_component`, `emplace_group_component`, `emplace_array_component`, and `emplace_nodegroup_component` should be used instead of `emplace_component`.  Note that `emplace_component` still works but will be removed in the future.
- The functions in Utilities/System/ParallelInfo now have two versions: the low-level charm++ version (in the `sys` namespace) and a new version templated on the `DistributedObject`.  The `sys` version should be used only for low-level things like `AbortWithErrorMessage` or the `ParallelInfo` executable where you know they will never be mocked; the templated version should be used everywhere else.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
Algorithm.hpp: Add wrappers to charm++ informational functions.

markscheel committed:
MockDistributedObject: Add mock charm++ informational functions.

markscheel committed:
ActionTesting: Add mocked nodes and cores.

MockRuntimeSystem now has a vector<size_t> passed into
its constructor; this specifies the number of (mock)
cores on each (mock) node.

Each MockDistributedObject and their proxies keep
track of the mock node and core that it lives on.

The various member function have been updated to do
the right thing for multiple nodes/cores.

There are now new functions emplace_singleton_component,
emplace_group_component, emplace_array_component,
and emplace_nodegroup_component.
emplace_component is deprecated but still exists until
everyone changes their unit tests.

markscheel committed:
Change some mock components to MockGroup and MockNodeGroup.

This is so that CkLocalBranch is called only from Groups and
Nodegroups in the ActionTesting framework.

markscheel committed:
Invoke.hpp: Allow a proxy to disable ckLocal.

This is used in ActionTesting.

markscheel committed:
Change some mock components to use Singletons.

This is so that ckLocal() is called only from Singletons.

markscheel committed:
Test_ActionTesting: TestNodesAndCores.

markscheel committed:
Clang-tidy.

markscheel committed:
Use emplace_(node)group_component in tests.

wthrowe commented on src/Parallel/Invoke.hpp:
```diff
@@ -32,6 +32,29 @@ struct has_ckLocal_method : std::false_type {};
+
+// Any class template that disables its ckLocal() function for some
+// template parameters (via static_assert or if constexpr) should
+// define a type alias ck_Local_disabled to be std::true_type for
```
`ckLocal_disabled` and again two lines down.

wthrowe commented on src/Parallel/Invoke.hpp:
```diff
@@ -32,6 +32,29 @@ struct has_ckLocal_method : std::false_type {};
+// where ckLocal() is enabled.  If ck_Local_disabled isn't defined,
+// then ckLocal() is understood to be enabled.
+template<typename T, typename = std::void_t<>>
+struct ckLocal_disabled {
```
Why not `ckLocal_enabled` to match the thing being implemented and avoid inverted logic?

markscheel replied:
Sure, I can call it `ckLocal_enabled`. Are you ok with structs that don't define `ckLocal_enabled` to be considered enabled?

wthrowe replied:
Yes.

wthrowe commented on tests/Unit/Framework/MockDistributedObject.hpp:
```diff
@@ -288,11 +303,20 @@ class MockDistributedObject {
   MockDistributedObject(
+      const NodeId node_id, const LocalCoreId local_core_id,
+      const std::vector<std::vector<size_t>>& mock_global_cores,
+      const std::vector<std::pair<size_t, size_t>>& mock_nodes_and_local_cores,
```
Take vectors by value and move.

wthrowe commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -72,9 +72,12 @@ struct get_array_index;
  * \snippet Test_TerminatePhase.cpp component
```
Commit message: (ActionTesting: Add mocked nodes and cores.)  "There are new functions..." should emplace_array_component be included?

wthrowe commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -518,37 +607,47 @@ class MockCollectionOfDistributedObjectsProxy {
-        std::move(args));
+    static_assert(std::is_same_v<ChareType, MockNodeGroupChare>,
+                  "Do not call threaded_action for other than a Nodegroup");
+    alg::for_each(*mock_distributed_objects_,
+                  [&args](auto& index_and_mock_distributed_object) noexcept {
+                    index_and_mock_distributed_object.second
+                        .template threaded_action<Action>(args);
+                  });
```
Why is this calling the action on every component?

markscheel replied:
Because this is a member function of MockCollectionOfDistributedObjectsProxy, i.e. a proxy to the entire chare, not to one element of the chare.  So just like simple_action, shouldn't threaded_action be a broadcast call to all elements?  If you want one element, you would use ckLocalBranch to get your local element (or index the MockCollectionOfDistributedObjectsProxy to get a different element if you want that), and that would return a MockDistributedObject (in the case of ckLocalBranch) or a proxy to one (in the case of indexing) that you could call threaded_action on for a single element.

Pls let me know if my understanding of threaded_action is incorrect.

wthrowe replied:
You can't run a threaded action on an entire component, and you can't even run it on a remote element.  I thought you couldn't even run it on a proxy, but our documentation and code disagree with each other on this:

docs/DevGuide/Parallelization.md:
> User controlled threading is possible by calling the non-entry method member function `threaded_action`.

src/Parallel/Algorithms/AlgorithmNodegroup.ci:
> template \<typename Action>
> entry void threaded_action();

@nilsdeppe which of these is correct?

markscheel replied:
@wthrowe If we aren't supposed to ever run a threaded action on an entire component, then `MockCollectionOfDistributedObjectsProxy::threaded_action` should just be removed; so that's what I did in a fixup commit. All the tests pass, so we never used that function anyway.

nilsdeppe replied:
The way we implement threaded entry methods they can be run on an entire component. I don't think we use that feature right now. I see no reason right now to disable it either. The threaded actions should behave (almost) entirely the same way as simple actions except that they don't lock the entire group.

wthrowe commented on tests/Unit/Framework/Tests/Test_ActionTesting.cpp:
```diff
@@ -542,4 +543,415 @@ SPECTRE_TEST_CASE("Unit.ActionTesting.MockComponent", "[Unit]") {
+
+struct ValueTag : db::SimpleTag {
+  using type = int;
+  static std::string name() noexcept { return "ValueTag"; }
```
Remove line.

markscheel commented:
Fixups to address review by @wthrowe

moxcodes commented on tests/Unit/Framework/MockDistributedObject.hpp:
```diff
@@ -114,6 +114,21 @@ using item_type_if_contained_t =
 }  // namespace detail
 
+/// Wraps a size_t representing the node number.  This is so the user
+/// can write things like emplace_component(NodeId{3},...)  instead of
```
add backticks to the code parts of the documentation.

markscheel commented:
Restored threaded_action.

wthrowe commented on tests/Unit/Framework/MockDistributedObject.hpp:
```diff
@@ -304,16 +304,16 @@ class MockDistributedObject {
-      const std::vector<std::vector<size_t>>& mock_global_cores,
-      const std::vector<std::pair<size_t, size_t>>& mock_nodes_and_local_cores,
+      const std::vector<std::vector<size_t>> mock_global_cores,
+      const std::vector<std::pair<size_t, size_t>> mock_nodes_and_local_cores,
```
No const.

markscheel replied:
oops. Yes, of course.

wthrowe reviewed (changes requested):
Squash directly.

markscheel commented:
Fixed MockDistributedObject and squashed.

moxcodes commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -568,5 +692,20 @@ struct get_array_index<ActionTesting::MockArrayChare> {
   using f = typename Component::array_index;
 };
+template <>
+struct get_array_index<ActionTesting::MockGroupChare> {
```
if it is the same for all possible specializations, why specify four times? I suppose this way it causes a compile-time error if something tries to use it on a non-mock-chare type, but you'd get a compile-time error when it tried to access the `::array_index` anyways.

markscheel replied:
Thinking about this more: Yes, I could avoid 4 specializations, but I don't see why you would get a compile-time error when trying to access `array_index`.  Couldn't someone specify in a Component a misspelled `using chare_type=MockArrayyyChare` and a correct `using array_index=size_t`, and then accessing array_index would work but there may be a more confusing compiler message elsewhere?

moxcodes replied:
Yeah, that's a reasonable point -- this probably does avoid some misbehavior earlier than it would if were just a generic template. I'm fine with it being left as-is.

moxcodes commented on tests/Unit/Framework/MockRuntimeSystem.hpp:
```diff
@@ -129,69 +134,235 @@ class MockRuntimeSystem {
+              size_t global_core = 0;
+              for (auto num_cores_on_this_node :
+                   number_of_mock_cores_on_each_mock_node) {
+                std::vector<size_t> global_cores{};
```
[optional] This inner loop can be simplified with `std::iota`. I felt like the outer loop could also be replaced with a sufficiently clever `std::inclusive_scan` but it wasn't actually shorter than the explicit loop.

moxcodes commented on tests/Unit/Framework/MockRuntimeSystem.hpp:
```diff
@@ -129,69 +134,235 @@ class MockRuntimeSystem {
+                .at(mock_global_cores_.at(node_id.value)
+                        .at(local_core_id.value))
+                .get(),
+            &(tuples::get<InboxesTag<Component>>(inboxes_)[array_index]),
```
Do I understand correctly that these instances of using `[array_index]` with the `inboxes_` works primarily because its an unordered map, and therefore this line is doing the work of both inserting the new element, then passing a pointer to that new element to the new `MockDistributedObject`?
That's a little confusing (and a reader is likely going to look around for where `inboxes_` got initialized and fail to find anything). Perhaps either a code comment explaining the trick, or use a `.insert` on a line before the `emplace` to separate out the two operations. 
(any edits should apply to the 8 similar `emplace` functions)

moxcodes commented on tests/Unit/Framework/Tests/Test_ActionTesting.cpp:
```diff
@@ -542,4 +543,414 @@ SPECTRE_TEST_CASE("Unit.ActionTesting.MockComponent", "[Unit]") {
+  using component_a = ComponentA<metavars>;
+
+  // Choose 2 nodes with 3 cores on first node and 2 cores on second node.
+  ActionTesting::MockRuntimeSystem<metavars> runner{{},{},{3,2}};
```
[nitpick] has this been formatted? I would have expected more space in the initializer list.

moxcodes commented on src/Utilities/System/ParallelInfo.hpp:
```diff
@@ -9,7 +9,7 @@
 #include <charm++.h>
 
-namespace sys {
+namespace sys::charm {
```
I'm confused... don't functions elsewhere use the functions in this namespace? Why didn't other things have to change?

moxcodes commented on src/Elliptic/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -83,7 +83,7 @@ void DgElementArray<Metavariables, PhaseDepActionList>::allocate_array(
         initial_element_ids(block.id(), initial_ref_levs);
     int which_proc = 0;
-    const int number_of_procs = sys::number_of_procs();
+    const int number_of_procs = sys::charm::number_of_procs();
```
Oh, I see..  This commit should be combined with the one that's the namespace change so everything still compiles on each commit.

moxcodes reviewed (commented):
I just have a few minor comments -- feel free to squash the changes to these.

markscheel commented:
Made changes suggested by @moxcodes including squashing the sys::charm namespace commits.  Squashed.

moxcodes commented:
Did something with git go wrong? I think the commits are still the same, including the separate namespace commits.

markscheel commented:
Yeah, something with git went wrong.  Need to redo because my local copy doesn't look like it should.

markscheel commented:
Made and squashed @moxcodes suggestions.  Also added a new fixup at request of @nilsleiffischer who wanted the node/core functions in Parallel/Info.hpp

nilsvu commented on src/Parallel/Info.hpp:
```diff
@@ -0,0 +1,101 @@
+// these functions rather than the raw charm++ versions (in the
+// sys::charm namespace below) so that the mocking framework will see
+// the mocked cores and nodes.
+namespace sys {
```
Could you make this the `Parallel` namespace and the one in `Utilities/System` just `sys` (instead of `sys::charm`)? Then if I understand correctly this PR keeps the low-level `sys` functions the same but introduces new `Parallel` functions that work with distributed objects.

markscheel replied:
I can put these into the `Parallel` namespace.  But I kind of like `sys::charm` for the other ones to emphasize that these are charm-dependent functions, and also to discourage people from using them (or make people think twice before using them).

One worry I have:  Suppose someone writes a new Action (or modifies an existing Action) and uses the charm functions. There will be no compiler error; the only effect will be that during ActionTesting the function will return the wrong answer (i.e. always one core and one node) and result in a bug that will be hard to find.  Thoughts?

nilsvu replied:
- Yea we can consider moving the "raw" functions from `sys` -> `sys::charm`, though I think that doesn't need to happen in this PR. I would prefer to keep them in `sys` because they are not inherently charm-specific (only their implementation calls into charm).
- I can't think of a foolproof way to make sure people use the `Parallel` functions, other than adding warnings to the docs. The `sys` functions have legitimate use-cases in layers of code that don't yet depend on `Parallel` (mostly error-handling) to avoid cyclic dependencies.
- The interface you suggested before, where the distributed-object functions are in `sys`, has the problem that distributed objects are a `Parallel` concept that shouldn't be lifted to `sys`. Since the functions are generic the cyclic dependency is technically broken (I think). However, the only argument I can see for putting the distributed-object functions in `sys` is to "hide" the raw functions a bit more by moving the `Parallel` functions closer to them, which doesn't really address the worry that people can use the raw functions.

So my suggestion is: Keep the raw functions in `sys`. Add functions to `Parallel` that work with distributed objects. Add warnings to the `sys` functions that they are super low-level and people should use the `Parallel` functions when they work with distributed objects. What do you think?

markscheel replied:
Ok, done!  Fixup commits pushed (and rebase onto current develop)

wthrowe commented on src/Parallel/Info.hpp:
```diff
@@ -0,0 +1,101 @@
+
+#pragma once
+
+// The functions in namespace `sys` that return information on nodes
+// and cores are templated on DistribObject.  Actions should use
+// these functions rather than the raw charm++ versions (in the
+// sys::charm namespace below) so that the mocking framework will see
+// the mocked cores and nodes.
```
Consider making this Doxygen.

nilsvu replied:
also there's a residual `sys::charm` -> `sys` in line 13

wthrowe commented on src/ParallelAlgorithms/LinearSolver/Observe.hpp:
```diff
@@ -18,7 +18,7 @@
 #include "Utilities/Functional.hpp"
 #include "Utilities/PrettyType.hpp"
-#include "Utilities/System/ParallelInfo.hpp"
+#include "Parallel/Info.hpp"
```
Alphabet.

wthrowe reviewed (commented):
I'm happy with the fixups.  My requests can be squashed directly.

nilsvu commented:
My concern in the thread https://github.com/sxs-collaboration/spectre/pull/2792#discussion_r568473959 is addressed. I think there's no need I jump in as another full reviewer, so no need to wait for my approval.

markscheel committed:
Add templated node/core functions in Parallel/Info.hpp

These are (sometimes) wrappers around the low-level functions
in Utilities/System/ParallelInfo.hpp, but they can be mocked
by the ActionTesting framework.

Add documentation telling users to use the templated node/core
functions instead of the low-level ones.

markscheel committed:
Use templated node/core calls in Actions.

markscheel commented:
Fixed the remaining things, squashed everything, rebased on develop.

kidder commented on tests/Unit/Framework/MockDistributedObject.hpp:
```diff
@@ -665,6 +721,13 @@ class MockDistributedObject {
+  size_t mock_node_{0};
+  size_t mock_local_core_{0};
+  // mock_global_cores[node][local_core] is the global_core.
+  std::vector<std::vector<size_t>> mock_global_cores_{};
```
how about `std::unordered_map<NodeId, std::unordered_map<LocalCoreId, GlobalCoreId>>`

kidder commented on tests/Unit/Framework/MockDistributedObject.hpp:
```diff
@@ -665,6 +721,13 @@ class MockDistributedObject {
+  // mock_global_cores[node][local_core] is the global_core.
+  std::vector<std::vector<size_t>> mock_global_cores_{};
+  // mock_nodes_and_local_cores_[global_core] is the pair node,local_core.
+  std::vector<std::pair<size_t, size_t>> mock_nodes_and_local_cores_{};
```
and here `std::unordered_map<GlobalCoreId, std::pair<NodeId,LocalCoreId>>`

kidder commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -485,17 +545,46 @@ class MockCollectionOfDistributedObjectsProxy {
+  using ckLocal_enabled = std::is_same<ChareType, MockSingletonChare>;
+
+  // ckLocal should be called only on a singleton.
+  MockDistributedObject<Component>* ckLocal() noexcept {
```
template this on   `template <typename U = ChareType, Requires<std::is_same_v<U, ChareType> and std::is_same_v<U, MockSingletonChare>>>` and then you shouldn't need `ckLocal_enabled`.

kidder commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -518,37 +607,47 @@ class MockCollectionOfDistributedObjectsProxy {
+  // mock_node_, mock_local_core_, and mock_global_core_ are the
+  // (mock) node and core that the
+  // MockCollectionOfDistributedObjectsProxy lives on.  This is
+  // different than the (mock) nodes and cores that each element ofthe
```
typo: of the

kidder commented on tests/Unit/Framework/MockRuntimeSystem.hpp:
```diff
@@ -129,69 +135,240 @@ class MockRuntimeSystem {
+        mutable_caches_(mock_nodes_and_local_cores_.size()),
+        caches_(mock_nodes_and_local_cores_.size()) {
+    // Fill the parallel components in each cache.  There is one cache
+    // per mock core.  The parallel components on each cache (i.e. the
```
on each core?

kidder commented on tests/Unit/Framework/MockRuntimeSystem.hpp:
```diff
@@ -129,69 +135,240 @@ class MockRuntimeSystem {
+      Options&&... opts) noexcept {
+    detail::get_initialization<Component>::initialize_databox_action::
+        set_initial_values(initial_values);
+    auto iterator_bool = mock_distributed_objects<Component>().emplace(
```
use a structured binding [iterator, a better_name_than_bool].   (Repeat below)

kidder commented on tests/Unit/Framework/MockRuntimeSystem.hpp:
```diff
@@ -203,6 +380,102 @@ class MockRuntimeSystem {
+        std::is_same_v<typename Component::chare_type, MockGroupChare>,
+        "emplace_group_component_and_initialize expects a MockGroupChare");
+    // Emplace once for each core, index by global_core.
+    for (const auto& node_core_pair : mock_nodes_and_local_cores_) {
```
structured binding [node, global_core]

kidder commented on tests/Unit/Framework/MockRuntimeSystem.hpp:
```diff
@@ -421,8 +700,18 @@ class MockRuntimeSystem {
-  Parallel::MutableGlobalCache<Metavariables> mutable_cache_{};
-  GlobalCache cache_{};
+  // mock_global_cores[node][local_core] is the global_core.
+  std::vector<std::vector<size_t>> mock_global_cores_{};
```
replace mock_global_cores_  and mock_nodes_and_local_cores_ with same types as commented on in MockDistributedObject

kidder commented on src/Parallel/Info.hpp:
```diff
@@ -0,0 +1,104 @@
+ * \brief Number of processing elements.
+ */
+template <typename DistribObject>
+int number_of_procs(const DistribObject& distributed_object) {
```
noexcept everything

kidder reviewed (commented):
Reviewed with @nilsdeppe.   Please consider a followup PR to address these minor changes that we think will improve usability
